File: core/bindings/java/com/mpiannucci/buoyfinder/core/Location.kt
package com.mpiannucci.buoyfinder.core

class Location {

    private val ptr: Long

    constructor(lat: Double, lon: Double, name: String) {
        ptr = new(lat, lon, name)
    }

    constructor(rawPtr: Long) {
        ptr = rawPtr
    }

    fun finalize() {
        free(ptr)
    }

    val latitude: Double
        get() = latitude(ptr)

    val longitude: Double
        get() = longitude(ptr)

    val altitude: Double
        get() = altitude(ptr)

    val name: String
        get() = name(ptr)

    private external fun new(lat: Double, lon: Double, name: String): Long
    private external fun free(ptr: Long)
    private external fun latitude(ptr: Long): Double
    private external fun longitude(ptr: Long): Double
    private external fun altitude(ptr: Long): Double
    private external fun name(ptr: Long): String
}

File: cargo/bindings/swift/Location.swift
//
//  Location.swift
//  BuoyFinder
//
//  Created by Matthew Iannucci on 9/20/18.
//  Copyright © 2018 Matthew Iannucci. All rights reserved.
//

import Foundation
import CoreLocation

class Location {
    public let raw: OpaquePointer
    
    init(ptr: OpaquePointer) {
        raw = ptr
    }
    
    init(lat: Double, lon: Double, name: String) {
        raw = location_new(lat, lon, name)
    }
    
    convenience init(location: CLLocationCoordinate2D) {
        self.init(lat: location.latitude, lon: location.longitude, name: "")
    }
    
    deinit {
        location_free(raw)
    }
    
    public var latitude: Double {
        get {
            return location_latitude(raw)
        }
    }
    
    public var longitude: Double {
        get {
            return location_longitude(raw)
        }
    }

    public var altitude: Double {
        get {
            return location_altitude(raw)
        }
    }
    
    public var name: String {
        get {
            return String(cString: location_name(raw))
        }
    }
    
    public var coreLocation: CLLocationCoordinate2D {
        get {
            return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
        }
    }
}
